### ISSUE TITLE
componentModel jsr330 not working with dagger DI

### ISSUE THREAD

**Issue by afdia** [mapstruct/mapstruct#571]
Currently if componentModel=jsr330, you're annotating the private fields with @Inject.

This works for most jsr330 based DI frameworks, but unfortunately not for dagger which is based on annotation processing and doesn't use reflection to access private fields (see https://google.github.io/dagger/)

Therefore to make the componentModel work with dagger, one of 2 things would be required:
1. Use constructor injection instead of field injection
-> then private fields can be injected without problems
2. Make fields package private instead
-> then the fields can be accessed without reflection

For dagger the first approach would be better, because they chose to require an @Inject annotated constructor for construction or otherwise the user has to create an appropriate @Provides method (see discussion at https://github.com/square/dagger/issues/412)

But the second approach would also work (and would be easier to implement I guess), even though the user has to make the injectMembers call himself.

And thanks for the nice mapping tool, it's good to have such powerful annotation processing based utilities around!

**Comment by agudian** [comment]
I'd prefer option one as well...

Question is... should we switch to constructor-injection for all jsr330 mappers? Or only if a certain option is set? Or should we add the component model "jsr330-constructor" as an alternative?

**Comment by gunnarmorling** [comment]
+1 For using constructor injection. I think we could switch to that in general, no option should be needed.

> And thanks for the nice mapping tool, it's good to have such powerful annotation processing based utilities around!

Thanks for the nice feedback! If you like, you can help by spreading the word on MapStruct by tweeting, blogging etc. :)

**Comment by afdia** [comment]
I'm not that active on social media, but I've already told several developer collegues about mapstruct ;)

I also think there should be no problem with switching jsr303 to constructor injection except some minor changes:
- users which subclass a Mapper implementation class would need to call the constructor to initialize the injected fields
- someone who is using an abstract class for the mappings and writes a constructor could expect that the impl class constructor calls the constructor (I don't know how MapStruct handles this case now)

On the positive side, the mapping impl classes would be easier to initialize in unit-tests which often do not depend on a DI framework even if the production code does

**Comment by megalobrainiac** [comment]
Constructor injection would also be very useful when using Spring DI.  While Spring does allow field injection adding (or switching) to constructor injection would allow for a better testing/mocking experience, especially when running unit tests outside of a Spring context.

**Comment by papo2608** [comment]
Are there any updates regarding constructor injection/using dagger?

**Comment by igor-dmitriev** [comment]
Hi guys, I am going to implement constructor-injection support, but we need to discuss it first. How can we do that? If we don't change basic implementation we can add a flag `useConstructorInjection=true` , like @agudian proposed

**Comment by gunnarmorling** [comment]
Hi @igor-dmitriev, awesome, that's great news! Does anything speaks against just always using constructor injection? If not, I'd prefer that. The less options and knobs, the better :)

Looking forward to the pull request, let us know if you run into an issues along the way. Thanks!

**Comment by igor-dmitriev** [comment]
@gunnarmorling cool, thanks, I agree, I prefer to use constructor-injection. I am wondering if it will be ok for all DI containers, cause I am just using Spring framework.

**Comment by filiphr** [comment]
@igor-dmitriev awesome for stepping up for this one. As @gunnarmorling, I also prefer always using constructor injection. The only thing where I see constructor injection making problems is cyclic dependency. At least for Spring it will fail to generate the Beans. 

In my opinion one should never introduce / use such cyclic dependency between beans, but it happens 😄. For this reason I think that we need to have a way to either use field or setter injection instead.

**Comment by gunnarmorling** [comment]
Ok, so Spring does not support constructor injection in case of cyclic dependencies between two beans? I thought they'd use proxies to resolve that. But if not, yes, then we should support both indeed.

**Comment by filiphr** [comment]
I just found out that there is a workaround to support it in the constructor. However, not without annotating the parameters with `@Lazy`. [This](http://www.baeldung.com/circular-dependencies-in-spring) is a nice article explaining all the workarounds. 

Do you know if other DI frameworks support injection of cyclic dependencies?

**Comment by gunnarmorling** [comment]
Not out of my head. I'd expect CDI to support it. Would be worth trying.

**Comment by afdia** [comment]
Typically for jsr330 you should be able to inject a Provider<Dep> instead of Dep and the provider instantiates the object after the first call of the get() method.
See also the [Guice wiki](https://github.com/google/guice/wiki/CyclicDependencies#break-the-cycle-with-a-provider).
The [javadoc from the jsr330 Provider class](https://javax-inject.github.io/javax-inject/api/javax/inject/Provider.html) mentions lazy injection or cyclic dependencies as a usecase

The problem is that I cannot call "new Mapper(new Dependency()) anymore but have to wrap the Dependency into a Provider interface.
Possible solutions:
- Generate both constructors (all dependencies directly passed and all dependencies passed passed with Providers) but annotate the one with Providers with `@Inject`
- add a switch (either between field and constructor injection or between direct and provider injection)

Some DI frameworks have proprietary ways to inject objects lazily (such as the link about Spring).
Dagger2 Lazy class is explained here: https://google.github.io/dagger/users-guide.html (section about lazy injection)
But the component model is called jsr330, therefore I guess it should stick with the standard.

**Comment by igor-dmitriev** [comment]
@filiphr you are right, I think we should give an opportunity to use setter/field/constructor injections. Let's make constructor injection by default and an injection flag should be present, line `injection=setter`

**Comment by agudian** [comment]
Switching to constructor injection by default might actually break things for current users. 

So'd actually prefer leaving the field injection the default and allowing to use constructor injection using an option.... Having those two, I don't see much benefit of also adding setter-injection -- in the long run it'll be just another thing we have to take care about.

**Comment by filiphr** [comment]
I think that eventually we need to switch to constructor injection by default. I would propose that either we use field injection in the Beta and the switch it in the final release or we announce the change and then we switch it in the release after we introduce this feature. 

Regarding the setter injection. Theoretically we could drop the field injection and use setter injection instead. Regarding Spring there are no difference about what you use. I am not sure about other frameworks though.

**Comment by kevcodez** [comment]
Any news on this? Would really like to see constructor injection, especially for testability.

**Comment by filiphr** [comment]
@kevcodez it is still pending, I am not sure if someone is working on it yet.

**Comment by kevcodez** [comment]
I submited a PR to support constructor injection: https://github.com/mapstruct/mapstruct/pull/1263

**Comment by filiphr** [comment]
Thanks a lot to @kevcodez for providing the PR for using constructor injection in the generated mappers. The PR #1263 has been merged into master and would be released with the next 1.3.

We have kept the old field injection as default (for now) and there is an option to enable constructor injection. It would be possible via:

* `@Mapper(injectionStrategy = InjectionStrategy.CONSTRUCTOR)`
* `@MapperConfig(injectionStrategy = InjectionStrategy.CONSTRUCTOR)`
